File: language/src/main/java/brj/nodes/builtins/PolyNode.java
package brj.nodes.builtins;

import brj.BridjeLanguage;
import brj.nodes.ExprNode;
import brj.runtime.BridjeContext;
import com.oracle.truffle.api.dsl.CachedContext;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.frame.VirtualFrame;

public abstract class PolyNode extends ExprNode {
    public PolyNode(BridjeLanguage lang) {
        super(lang);
    }

    @Specialization
    public Object doExecute(VirtualFrame frame,
                            @CachedContext(BridjeLanguage.class) BridjeContext ctx) {
        return ctx.poly((String) frame.getArguments()[1], ((String) frame.getArguments()[2]));
    }
}

File: language/src/main/kotlin/brj/builtins/PolyNode.kt
package brj.builtins

import brj.BridjeLanguage
import com.oracle.truffle.api.TruffleLanguage.ContextReference
import com.oracle.truffle.api.dsl.Specialization

private val CTX_REF = ContextReference.create(BridjeLanguage::class.java)

@BuiltIn("poly")
abstract class PolyNode(lang: BridjeLanguage) : BuiltInFn(lang) {
    @Specialization
    fun doPoly(lang: String, code: String) =
        CTX_REF[this].poly(lang, code)
}